apps/src/p5lab/spritelab/CoreLibrary.js

```diff
@@ -345,19 +352,56 @@ export default class CoreLibrary {
     return spriteIds;
   }
 
+  getNumberOfSprites() {
+    return Object.keys(this.nativeSpriteMap).length;
+  }
+
+  getCappedNumSprites(numRequested) {
+    const numSpritesSoFar = this.getNumberOfSprites();
+    // (MAX_NUM_SPRITES + 1) is the actual maximum number of sprites possible
+    // At MAX_NUM_SPRITES, the workspace alert warning is displayed
+    const numNewSpritesPossible = MAX_NUM_SPRITES + 1 - numSpritesSoFar;
+    return Math.min(numRequested, numNewSpritesPossible);
+  }
+
   getLastSpeechBubbleForSpriteId(spriteId) {
     const speechBubbles = this.speechBubbles.filter(
       ({sprite}) => sprite.id === parseInt(spriteId)
     );
     return speechBubbles[speechBubbles.length - 1];
   }
 
+  // When the total number of sprites reaches MAX_NUM_SPRITES,
+  // a workspace alert warning is displayed.
+  // (MAX_NUM_SPRITES + 1) is the actual maximum number of sprites possible
+  // and once this number is reached, this function returns true
+  reachedSpriteLimit() {
+    const numSprites = this.getNumberOfSprites();
+    if (numSprites >= MAX_NUM_SPRITES + 1) {
```

Comment:
small nit :) 

numSprites > MAX_NUM_SPRITES 

In my head is more human readable and matches your method definition without an operation going through the devs head.

Reply:
sorry, my bad, 

I get confused with the definition of Max_NUM_SPRITES which is not the real max 
`(MAX_NUM_SPRITES + 1) is the actual maximum number of sprites possible`

Reply:
Can a new variable be created 

MAX_NUM_SPRITES = 1001;
MAX_NUM_SPRITE_WARNING = MAX_NUMBER_SPRITES - 1;

And then it would avoid confusions when reading the code? We could even add some comments around the second variable to understand more about it. I can see someone forgetting a +1 somewhere and introducing a bug

Reply:
I LOVE this idea of using two constants. I was struggling with which one to use, but much better to use 2 different constants! And I like the name MAX_NUM_SPRITE_WARNING - will use. Thanks so much!